tests/mir_test_framework/window_management_test_harness.cpp

```diff
@@ -0,0 +1,213 @@
+ */
+
+#include "mir_test_framework/window_management_test_harness.h"
+#include <mir/executor.h>
+#include <mir/scene/surface.h>
+#include <mir/shell/surface_specification.h>
+#include <mir/wayland/weak.h>
+#include <miral/window_manager_tools.h>
+#include <miral/minimal_window_manager.h>
+#include <miral/set_window_management_policy.h>
+#include "mir/test/doubles/fake_display.h"
+#include "mir/test/doubles/stub_buffer_stream.h"
+#include "src/miral/basic_window_manager.h"
+
+#include <queue>
+
+namespace ms = mir::scene;
+namespace msh = mir::shell;
+namespace mf = mir::frontend;
+namespace mi = mir::input;
+namespace mtd = mir::test::doubles;
+namespace mg = mir::graphics;
+using namespace testing;
+
+namespace
+{
+class WorkQueue
+{
+public:
+    void enqueue(std::function<void()> const& f)
+    {
+        work_queue.push(f);
+    }
+
+    auto next() -> bool
+    {
+        if (work_queue.empty())
+            return false;
+
+        auto const& f = work_queue.front();
+        f();
+        work_queue.pop();
+        return !work_queue.empty();
+    }
+
+    void run()
+    {
+        while (next()) {}
+    }
+
+private:
+    std::queue<std::function<void()>> work_queue;
+};
+
+}
+
+struct mir_test_framework::WindowManagementTestHarness::Self : public ms::SurfaceObserver
+{
+    explicit Self(WindowManagementTestHarness* harness)
+         : harness{harness}
+    {
+    }
+
+    void attrib_changed(mir::scene::Surface const*, MirWindowAttrib, int) override {}
+    void window_resized_to(mir::scene::Surface const*, mir::geometry::Size const&) override {}
+    void content_resized_to(mir::scene::Surface const*, mir::geometry::Size const&) override {}
+    void moved_to(mir::scene::Surface const*, mir::geometry::Point const&) override {}
+    void hidden_set_to(mir::scene::Surface const*, bool) override {}
+    void frame_posted(mir::scene::Surface const*, mir::geometry::Rectangle const&) override {}
+    void alpha_set_to(mir::scene::Surface const*, float) override {}
+    void orientation_set_to(mir::scene::Surface const*, MirOrientation) override {}
+    void transformation_set_to(mir::scene::Surface const*, glm::mat4 const&) override {}
+    void reception_mode_set_to(mir::scene::Surface const*, mir::input::InputReceptionMode) override {}
+
+    void cursor_image_set_to(
+        mir::scene::Surface const*,
+        std::weak_ptr<mir::graphics::CursorImage> const&) override {}
+
+    void client_surface_close_requested(mir::scene::Surface const* surf) override
+    {
+        auto it = std::find_if(known_surfaces.begin(), known_surfaces.end(), [surf](std::shared_ptr<ms::Surface> const& s)
+        {
+            return s.get() == surf;
+        });
+
+        if (it != known_surfaces.end())
+        {
+            work_queue.enqueue([shell=harness->server.the_shell(), surface=*it]()
+            {
+                shell->destroy_surface(surface->session().lock(), surface);
+            });
+        }
```

Comment:
It's a pity this can't be just
```
shell->destroy_surface(surf->session().lock(), surf)
```
but
1. That would require changing the `destroy_surface` signature, which is out-of-scope here, and
2. I'm not sure that dropping the `shared_ptr` from the `destroy_surface` signature would be safe anyway.